### ISSUE TITLE
DUAL-SNOW produces different void network, including spurious pore throats

### ISSUE THREAD

**Issue by kweishaupt** [PMEAL/porespy#326]
I have tried to extract a dual network from RAW data published here: https://www.digitalrocksportal.org/projects/11/origin_data/17/

I used ImageJ to crop the RAW data to 500^3 voxels, starting from (0,0,0) and saved the file as `nrrd` format.

Then I used the followign python script to run `poreSpy`(latest release):

```python
# import necessary packages
import openpnm as op
import porespy as ps
import nrrd
import numpy as np

filename_input = 'cropped.nrrd';
filename_output = 'bentheimer_dual_snow_nrrd';

# read inputfile
im, header = nrrd.read(filename_input);
im = ~np.array(im, dtype=bool)

# print porosity
porosity = ps.metrics.porosity(im);
print(porosity);


#application of snow-algorithm
#snow_output = ps.networks.snow_dual(im,
                 # voxel_size=3.0035e-6,
                 # boundary_faces=[],
                #  marching_cubes_area=False)

snow_output = ps.networks.snow(im,
                  voxel_size=3.0035e-6,
                  boundary_faces=[],
                  marching_cubes_area=False)

#create empty dummy network
pn = op.network.GenericNetwork(project=project);
#update openpnm pore network
pn.update(snow_output);
#trimming pore network to avoid singularity
print('Number of pores before trimming: ', pn.Np);
h = pn.check_network_health();
op.topotools.trim(network=pn, pores=h['trim_pores']);
print('Number of pores after trimming: ', pn.Np);
#export to vtk
op.io.VTK.save(network=pn, filename=filename_output)

```
Using `ps.networks.snow` yields a good-looking network:

![image](https://user-images.githubusercontent.com/74908731/99970772-99001680-2d9c-11eb-885a-5059b57736d2.png)

However, `ps.networks.snow_dual` yields almost the same void network but one of the pores seems to have a lot of
spurious connections (black is the SNOW network, red is the DUAL-SNOW network with a ParaView threshold filter showing only the void throats).
![image](https://user-images.githubusercontent.com/74908731/99971134-12980480-2d9d-11eb-87de-1b9c994d9f18.png)

Is there anything I could do to improve the result? Did I do something wrong or could it be a bug?

__Extra Question__: 
Doing the same procedure as above for SNOW using a `tiff` or `RAW` input file will also mess up the generated grid (in a different way).
I followed the advice given [here](https://github.com/PMEAL/OpenPNM/issues/1373#issuecomment-578215692) to load the raw image. I can also open a separate issue for that.

**Comment by Zohaib-Atiq** [comment]
@kweishaupt I re-run your code with .tif format and It seems to work fine for both `SNOW` and `SNOW_DUAL`. May be there is something wrong in how you are reading the image. For your help I am attaching my script so that you can have a look why things are wrong for you. 
```
import openpnm as op
import porespy as ps
import numpy as np
import imageio as io


filename_input = 'Bentheimer_1000c_3p0035um.tif'
filename_output = 'bentheimer_dual_snow_nrrd'

# read inputfile
im = io.mvolread(filename_input, format='tif')[0]

im = im[0:500, 0:500, 0:500]
im = ~np.array(im, dtype=bool)

# print porosity
porosity = ps.metrics.porosity(im)
print(porosity)


#application of snow-algorithm
snow_output = ps.networks.snow_dual(im,
                 voxel_size=3.0035e-6,
                 boundary_faces=[],
                 marching_cubes_area=False)

# snow_output = ps.networks.snow(im,
#                   voxel_size=3.0035e-6,
#                   boundary_faces=[],
#                   marching_cubes_area=False)

#create empty dummy network
pn = op.network.GenericNetwork()
#update openpnm pore network
pn.update(snow_output)
#trimming pore network to avoid singularity
print('Number of pores before trimming: ', pn.Np)
h = pn.check_network_health()
op.topotools.trim(network=pn, pores=h['trim_pores'])
print('Number of pores after trimming: ', pn.Np)
#export to vtk
op.io.VTK.save(network=pn, filename=filename_output)
```

SNOW OUTPUT:
![SNOW](https://user-images.githubusercontent.com/30351879/99997700-892c0680-2d8b-11eb-84d7-bc26d97cf12c.PNG)

SNOW DUAL OUTPUT
![SNOW_DUAL](https://user-images.githubusercontent.com/30351879/99997853-bf698600-2d8b-11eb-887a-329d5d2dc367.PNG)

**Comment by ma-sadeghi** [comment]
@Zohaib-Atiq @kweishaupt Maybe it has to do with your `porespy` version. Please use the `dev` branch. Here's how you install it:

```bash
git clone https://github.com/PMEAL/porespy
cd porespy
pip install -e .
```

Make sure to remove your previous `porespy`. 

PS. we recently addressed (PR #316) a bug in `porespy` where `SNOW` generated some spurious connections.

**Comment by kweishaupt** [comment]
@Zohaib-Atiq @ma-sadeghi Thanks a lot for your super fast response!

I  removed `openPNM` and `poreSpy` using `pip uninstall` and re-installed the `git dev` versions instead.
Unfortunately, the problem persists, even after copy and pasting @Zohaib-Atiq's code.

My guess would be that I somehow messed up  cropping  the raw data.

I'm not familiar with ImageJ (or Fiji), so here is what I did on Ubuntu 20.04

* download and extract the raw files from https://www.digitalrocksportal.org/projects/11/origin_data/17/
* start Fiji (ImageJ 1.53c)
* File -> Open -> chose "bentheimer.raw"
* Image type: 8 bit, Width: 1000 pixels, Height 1000 pixels, Offset to first Image: 0 bytes, Number of images: 1000, Gap between images: 0 bytes, check "Little-endian byte order" -> OK
* Edit -> Selection -> Specify
* Width: 500, Height: 500, X coordinate: 0, Y coordinate: 0, Slice: 0 -> OK
* Image -> Crop
* Image -> Stacks -> Tools -> Slice Keeper
* First Slice: 1, Last Slice: 500, Increment: 1
* File -> Save as -> {Tiff / nrrd / ... }

This will result in two different `vtp` files after network extractions with `snow(...)`

For `tif`:
```python
im = io.mvolread(filename_input, format='tif')[0]
im = im[0:500, 0:500, 0:500]
im = ~np.array(im, dtype=bool)
```
For `nrrd`
```python
import nrrd
im, header = nrrd.read(filename_input);
im = ~np.array(im, dtype=bool)
```

Red is based on the `tif` file, blue is based on the `nrrd` file (and matches substantially better with the 3D image, as shown above).

![image](https://user-images.githubusercontent.com/74908731/100010323-6243f380-2dd0-11eb-974b-ec2a25e73463.png)

Both `tif` and `nrrd` yield a broken `dual_snow` network.

**Comment by Zohaib-Atiq** [comment]
@kweishaupt Can you use this script and check If you are getting desired results:

```
import openpnm as op
import porespy as ps
import numpy as np

filename_output = 'bentheimer_dual_snow_nrrd2'

# read inputfile
im = np.fromfile("D:\\bentheimer.raw", dtype='uint8', sep="").reshape(1000, 1000, 1000)

#Crop Image
im = im[0:200, 0:200, 0:200]
im = ~np.array(im, dtype=bool)

# print porosity
porosity = ps.metrics.porosity(im)
print(porosity)

#application of snow-algorithm
snow_output = ps.networks.snow_n(im + 1,
                 voxel_size=3.0035e-6,
                 boundary_faces=None,
                 marching_cubes_area=False)

# snow_output = ps.networks.snow(im,
#                   voxel_size=3.0035e-6,
#                   boundary_faces=None,
#                   marching_cubes_area=False)

#create empty dummy network
pn = op.network.GenericNetwork()
#update openpnm pore network
pn.update(snow_output)
#trimming pore network to avoid singularity
print('Number of pores before trimming: ', pn.Np)
h = pn.check_network_health()
op.topotools.trim(network=pn, pores=h['trim_pores'])
print('Number of pores after trimming: ', pn.Np)
#export to vtk
op.io.VTK.save(network=pn, filename=filename_output)
```

**Comment by Zohaib-Atiq** [comment]
@kweishaupt I have reproduced the bug you were getting in `SNOW_Dual`. I have pushed a fix for it. You can use `snow_n` in above code for now to extract dual network which will give you the same network. 
Once this fix is merged into `dev` you can use SNOW_Dual as well for your image.

**Comment by kweishaupt** [comment]
@Zohaib-Atiq Thanks a lot for your suggestion with `snow_n` and thanks for the fix! That does work nicely in combination with the `nrrd` file.

There is still an issue with the difference between `raw` (or `tif`, red, cropped as described in your python code) and `nrrd` (blue, file created by ImageJ), which probably has nothing to do with `SNOW_DUAL` as it also occurs when only using `SNOW`

![image](https://user-images.githubusercontent.com/74908731/100073984-79b9c580-2e3e-11eb-92c4-a927c5d30875.png)

As you can see, the network is somehow wrongly rotated (?) and doesn't fit with the 3d data.

I've also seen that

```python
# load raw image, reshape and crop
im_raw = np.fromfile("bentheimer.raw", dtype='uint8', sep="").reshape(1000, 1000, 1000)
im_raw = im_raw[0:500, 0:500, 0:500]

# load nrrd image cropped by ImageJ
im_nrrd = nrrd.read("bentheimer_cropped.nrrd")[0]

assert((im_raw == im_nrrd).all())
```

fails. So there seems to be some different ordering which I could not figure out yet.


__Edit__:

I found the difference between `raw` and `nrrd`

```python
im_raw = np.swapaxes(im_raw,  0, 2)
assert((im_raw == im_nrrd).all())
```

now passes.

**Comment by kweishaupt** [comment]
@jgostick @Zohaib-Atiq I think I stumbled across another case where even `snow_n` yields a different void network than `snow`.
If you want, I can send you the binary image and the script I used.

**Comment by jgostick** [comment]
Cripes, sometimes it feels like the bug never end.  Please send the info.  Thanks!

**Comment by kweishaupt** [comment]
[test_berea.zip](https://github.com/PMEAL/porespy/files/5868394/test_berea.zip)
Here are  the data.

I first extracted a "normal" network using

``` python  run_snow.py Berea_center_cropped_200_200.raw -r 5.345e-6 -v 200 200 200```

Then I extracted a dual grid with 
``` python run_snow.py Berea_center_cropped_200_200.raw -r 5.345e-6 -v 200 200 200 -d```

Here is the result. SNOW, on the left, produces a good result.

SNOW_DUAL (`snow_n`  or `snow_dual` on the branch `snow_dual_bug`) features some extra throats (marked purple) on the lower right.

I first tried to manually deleted them but then I noticed that also the pore properties (such as diameter shown in the pics) differ.


![image](https://user-images.githubusercontent.com/74908731/105744689-c2285900-5f3d-11eb-89bf-73008dcd2a65.png)

Here is a close-up of the extra throats with the 3D void grid shown as well They appear to be in a large solid inclusion:
![image](https://user-images.githubusercontent.com/74908731/105745653-cacd5f00-5f3e-11eb-90d1-a36d44a4ccb5.png)

**Comment by kweishaupt** [comment]
I think the "extra throats" get removed by  

```python
h = pn.check_network_health()
op.topotools.trim(network=pn, pores=h['trim_pores'])
```
For dual snow, they are not really "insular" because they have interconnect neighbors. However, I think they should not even appear at that position in the first place.

 What really confuses me is the different set of properties:
`pore.extended_diameter` and `throat.length` differ while the remaining properties seem to be the same at first glance.

**Comment by kweishaupt** [comment]
@jgostick Thanks for fixing the discrepancy between `snow_dual` and `snow_n`. However, I think the issue described above still persists. Sorry!

**Comment by jgostick** [comment]
I merged a PR yesterday that was set to automatically close this issues, I didn't mean to close it.  I'll keep digging in to this.  Renovating porespy has moved to top priority now.

**Comment by jgostick** [comment]
I think I have found another bug that may be responsible for this.  For some reason the ``snow_partitioning`` function does not actually label ALL the void space.  There will be small pockets (usually isolated) that do get a region label.  When the additional watersheds (for phase 2+) are added to the previous ones, these pockets are are given a region number when we bump the region numbers of the 2nd watershed to be the max of the original + 1.  

Do the spurious throats by any chance connect  to pore N, where is the number of regions in the first watershed?  This would be the symptom.  

I have done a HUGE renovation of the snow/snow_dual/snow_n functions that addresses this, so will be fixed in V2.  Please let me know if this explains your observations.

**Comment by kweishaupt** [comment]
@jgostick  I just tried the most recent version of PoreSpy but it seems there are still some spurious void phase throats:

![image](https://user-images.githubusercontent.com/74908731/117688544-27d0e600-b1b9-11eb-802e-a4ffc55637ea.png)

I didn't have time to investigate this in more detail (i.e., have a look at the individual segmentation steps) but if you
want to reproduce this (using [this data](https://github.com/PMEAL/porespy/files/5868394/test_berea.zip)), here is the updated script:

```
import openpnm as op
import porespy as ps
import numpy as np
import argparse

parser = argparse.ArgumentParser(description='extract network from image using PoreSpy')
parser.add_argument('file', type=str, help='The image file')
parser.add_argument('-r', '--resolution', type=float, help='The voxel size', required=True)
parser.add_argument('-v', '--voxels', nargs='+', type=int, help='The number of voxels in each direction', default=[])
parser.add_argument('-n', '--outputname', type=str, help='The output file name', default='')
parser.add_argument('-m','--marchingCubesArea', action='store_true', help="Use marching cubes alg. to calculate throat area")
parser.add_argument('-d','--dualSnow', action='store_true', help="Extract dual network")
args = vars(parser.parse_args())

if args['file'].endswith('raw'):
    if len(args['voxels']) != 3:
        raise IOError('argument --voxels numX numY numZ must be given')
    print("Reading raw image file", args['file'], "with", args['voxels'], "voxels")
    if not args['outputname']:
        args['outputname'] = args['file'].replace('.raw', '')
    im = np.fromfile(args['file'], dtype='uint8', sep="").reshape(args['voxels'])
    im = ~np.array(im, dtype=bool)
    im = np.swapaxes(im, 0, 2)

elif args['file'].endswith('nrrd'):
    import nrrd
    im, header = nrrd.read(args['file'])
    im = ~np.array(im, dtype=bool)
    args['voxels'] = header['sizes']
    print("Reading nrrd image file", args['file'], "with", args['voxels'], "voxels")
    if not args['outputname']:
        args['outputname'] = args['file'].replace('.nrrd', '')

else:
    raise IOError('Invalid file format')


print("Porosity is", ps.metrics.porosity(im))

if args['marchingCubesArea']:
    print('Using Marching Cubes Alg. to calculate throat areas')

#application of snow-algorithm
if not args['dualSnow']:
    snowOutput = ps.networks.snow2(im,
                                   voxel_size=args['resolution'],
                                   accuracy=('high' if args['marchingCubesArea'] else 'standard'))
else:
    snowOutput = ps.networks.snow2(im+1,
                                   voxel_size=args['resolution'],
                                   accuracy=('high' if args['marchingCubesArea'] else 'standard'),
                                   phase_alias={1: 'solid', 2: 'void'})

# use porespy to sanitize some parameters (might become obsolete as some point)
pn, geo = op.io.PoreSpy.import_data(snowOutput.network)

# trimming pore network to avoid singularity
print('Number of pores before trimming: ', pn.Np)
h = pn.check_network_health()
op.topotools.trim(network=pn, pores=h['trim_pores'])
print('Number of pores after trimming: ', pn.Np)

filename = filename=args['outputname'] + ('_dual' if args['dualSnow'] else '')

#export to vtk
op.io.VTK.save(network=pn, filename=filename)
print("Writing", filename + ".vtp")

# export network for openpnm
pn.project.save_project(filename)
print("Writing", filename + ".pnm")
```

**Comment by jgostick** [comment]
I cannot reproduce this.  I have written an extensive unit test, and it checks for the number of pores and throats across several different permutations of the input arguments, and everything adds up. What's odd though, is that when I visualize the network in paraview I get neither of your images above.  It's a hybrid of both cases.  But I get the same image regardless of whether I do single phase, dual phase, with or without BC, serial or parallel.  So I'm stumped.

**Comment by jgostick** [comment]
Ahh, I think I see.  You trim the disconnected pores.  When single phase, the 'spurious' throats get trimmed, but when dual phase the are kept because they are actually connected to solid phase so as far as openpnm is concerned they are not disconnected.  Then when you go into paraview and threshold by the pore.void, they appear as spurious, but this is only because they were deleted from the void-only network, but not from the dual phase.  I'm satisfied that this is solved now, so closing.
